File: leetcode/array/reshape-the-matrix.cpp
class Solution {
public:
  vector<vector<int>> matrixReshape(vector<vector<int>> &mat, int r, int c) {
    int m = size(mat), n = size(mat[0]), total = m * n;
    if (r * c != total)
      return mat;

    vector<vector<int>> s(r, vector<int>(c));

    for (int i = 0; i < total; i++) {
      s[i / c][i % c] = mat[i / n][i % n];
    }

    return s;
  }
};

File: leetcode/array/reshape-the-matrix.go
func matrixReshape(mat [][]int, r int, c int) [][]int {
	m, n := len(mat), len(mat[0])
	total := m * n
	if r*c != total {
		return mat
	}

	res := make([][]int, r)
	for i := range res {
		res[i] = make([]int, c)
	}

	for i := 0; i < total; i++ {
		res[i/c][i%c] = mat[i/n][i%n]
	}
	return res
}
